File: src/lossy_state_graph.py
#!/usr/bin/env python3

import sys
from random import randrange
from pprint import pformat
from state_graph import StateGraph, Node

class LossyStateGraph(StateGraph):

    def __init__(self, capacity, window_size):
        super().__init__(capacity)

        self.drifted_tree_counter = 0
        self.window_size = window_size

    def get_next_tree_id(self, src):
        cur_node = self.graph[src]
        if not cur_node or cur_node.total_weight == 0:
            return -1

        r = randrange(cur_node.total_weight)
        cur_sum = 0

        for key, val in cur_node.neighbors.items():
            cur_sum += val[0]
            if r < cur_sum:
                val[1] += 1
                return key

        return -1

    def update(self, warning_tree_count):
        self.drifted_tree_counter += warning_tree_count
        if self.drifted_tree_counter < self.window_size:
            return

        self.drifted_tree_counter -= self.window_size

        # lossy count
        for node in self.graph:
            if node is None:
                continue

            for nei_key, val in list(node.neighbors.items()):
                # increment freq by #hits
                val[0] += val[1]
                node.total_weight += val[1]

                # reset #hits
                val[1] = 0

                # decrement freq by 1
                val[0] -= 1
                node.total_weight -= 1

                if val[0] <= 0:
                    # remove edge
                    self.graph[nei_key].indegree -= 1
                    self.__try_remove_node(nei_key)

                    del node.neighbors[nei_key]

            self.__try_remove_node(node.key)

    def __try_remove_node(self, key):
        node = self.graph[key]

        if node.indegree == 0 and len(node.neighbors) == 0:
            self.graph[key] = None

    def add_node(self, key):
        self.graph[key] = LossyNode(key)

    def add_edge(self, src, dest):
        if self.graph[src] == None:
            self.add_node(src)
        if self.graph[dest] == None:
            self.add_node(dest)

        src_node = self.graph[src]

        if dest not in src_node.neighbors.keys():
            src_node.neighbors[dest] = [0, 0]
            self.graph[dest].indegree += 1

        # update #hits
        src_node.neighbors[dest][1] += 1


class LossyNode(Node):
    def __init__(self, key):
        super().__init__(key)
        self.indegree = 0
        # self.neighbors = dict() # <tree_id, [weight, num_hit]>

File: src/cpp/lossy_state_graph.cpp
#include "lossy_state_graph.h" 

lossy_state_graph::lossy_state_graph(int capacity,
                                     int window_size) 
        : capacity(capacity), 
          window_size(window_size) {

    is_stable = false;
    graph = vector<unique_ptr<node_t>>(capacity);
}

int lossy_state_graph::get_next_tree_id(int src) {
    if (!graph[src] || graph[src]->total_weight == 0) {
        return -1;
    }

    int r = rand() % graph[src]->total_weight;
    int sum = 0;

    // weighted selection
    for (auto nei : graph[src]->neighbors) {
        sum += nei.second;
        if (r < sum) {
            graph[src]->neighbors[nei.first]++;
            graph[src]->total_weight++;
            return nei.first;
        }
    }

    return -1;
}

bool lossy_state_graph::update(int warning_tree_count) {
    drifted_tree_counter += warning_tree_count;

    if (drifted_tree_counter < window_size) {
        return false;
    }

    drifted_tree_counter -= window_size;

    // lossy count
    for (int i = 0; i < graph.size(); i++) {
        if (!graph[i]) {
            continue;
        }

        for (auto& nei : graph[i]->neighbors) {
            // decrement freq by 1
            graph[i]->total_weight--;
            nei.second--; // freq

            if (nei.second == 0) {
                // remove edge
                graph[nei.first]->indegree--;
                try_remove_node(nei.first);

                graph[i]->neighbors.erase(nei.first);
            }
        }

        try_remove_node(i);
    }

    return true;
}

void lossy_state_graph::try_remove_node(int key) {
    if (graph[key]->indegree == 0 && graph[key]->neighbors.size() == 0) {
        graph[key].reset();
    }
}

void lossy_state_graph::add_node(int key) {
    if (key >= capacity) {
        cout << "id exceeded graph capacity" << endl;
        return;
    }

    graph[key] = make_unique<node_t>();
}

void lossy_state_graph::add_edge(int src, int dest) {
    if (!graph[src]) {
        add_node(src);
    }

    if (!graph[dest]) {
        add_node(dest);
    }

    graph[src]->total_weight++;

    if (graph[src]->neighbors.find(dest) == graph[src]->neighbors.end()) {
        graph[src]->neighbors[dest] = 0;
        graph[dest]->indegree++;
    }

    graph[src]->neighbors[dest]++;
}

void lossy_state_graph::set_is_stable(bool is_stable_) {
    is_stable = is_stable_;
}

bool lossy_state_graph::get_is_stable() {
    return is_stable;
}
